File: shared_resources/src/main/java/com/thewizrd/shared_resources/tzdb/TZDBCache.java
package com.thewizrd.shared_resources.tzdb;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.room.Room;
import androidx.room.migration.Migration;
import androidx.sqlite.db.SupportSQLiteDatabase;

import com.thewizrd.shared_resources.SimpleLibrary;
import com.thewizrd.shared_resources.database.TZDatabase;
import com.thewizrd.shared_resources.tasks.AsyncTask;
import com.thewizrd.shared_resources.utils.AnalyticsLogger;
import com.thewizrd.shared_resources.utils.StringUtils;

import java.util.concurrent.Callable;

public class TZDBCache {
    private static TZDatabase tzDB;

    private static final Migration TZDB_MIGRATION_0_5 = new Migration(0, 5) {
        @Override
        public void migrate(@NonNull SupportSQLiteDatabase database) {
            // Since we didn't alter the table, there's nothing else to do here.
        }
    };

    private static final Migration TZDB_MIGRATION_5_6 = new Migration(5, 6) {
        @Override
        public void migrate(@NonNull SupportSQLiteDatabase database) {
            // Since we didn't alter the table, there's nothing else to do here.
        }
    };

    private static final Migration TZDB_MIGRATION_6_7 = new Migration(6, 7) {
        @Override
        public void migrate(@NonNull SupportSQLiteDatabase database) {
            // Since we didn't alter the table, there's nothing else to do here.
        }
    };

    private static final Migration[] TZDB_MIGRATION_SET = new Migration[]{
            TZDB_MIGRATION_0_5, TZDB_MIGRATION_5_6, TZDB_MIGRATION_6_7
    };

    public static String getTimeZone(final double latitude, final double longitude) {
        if (latitude != 0 && longitude != 0) {
            AnalyticsLogger.logEvent("TZDBCache: querying");
            // Initialize db if it hasn't been already
            if (tzDB == null) {
                Context context = SimpleLibrary.getInstance().getAppContext();
                tzDB = Room.databaseBuilder(context,
                        TZDatabase.class, "tzdb.db")
                        .addMigrations(TZDB_MIGRATION_SET)
                        .fallbackToDestructiveMigration()
                        .build();
            }

            // Search db if result already exists
            final String dbResult = AsyncTask.await(new Callable<String>() {
                @Override
                public String call() {
                    return tzDB.tzdbDAO().getTimeZoneData(latitude, longitude);
                }
            });

            if (!StringUtils.isNullOrWhitespace(dbResult))
                return dbResult;

            // Search tz lookup
            final String result = AsyncTask.await(new Callable<String>() {
                @Override
                public String call() {
                    return new TimeZoneProvider().getTimeZone(latitude, longitude);
                }
            });

            if (!StringUtils.isNullOrWhitespace(result)) {
                // Cache result
                AsyncTask.run(new Runnable() {
                    @Override
                    public void run() {
                        TZDB tzdb = new TZDB(latitude, longitude, result);
                        tzDB.tzdbDAO().insertTZData(tzdb);

                        // Run GC since tz lookup takes up a good chunk of memory
                        System.gc();
                    }
                });
            }

            return result;
        }

        return "UTC";
    }
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/tzdb/TZDBCache.kt
package com.thewizrd.shared_resources.tzdb

import androidx.room.Room
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import com.thewizrd.shared_resources.SimpleLibrary
import com.thewizrd.shared_resources.database.TZDatabase
import com.thewizrd.shared_resources.utils.AnalyticsLogger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

object TZDBCache {
    private var tzDB: TZDatabase? = null

    private val TZDB_MIGRATION_0_5: Migration = object : Migration(0, 5) {
        override fun migrate(database: SupportSQLiteDatabase) {
            // Since we didn't alter the table, there's nothing else to do here.
        }
    }

    private val TZDB_MIGRATION_5_6: Migration = object : Migration(5, 6) {
        override fun migrate(database: SupportSQLiteDatabase) {
            // Since we didn't alter the table, there's nothing else to do here.
        }
    }

    private val TZDB_MIGRATION_6_7: Migration = object : Migration(6, 7) {
        override fun migrate(database: SupportSQLiteDatabase) {
            // Since we didn't alter the table, there's nothing else to do here.
        }
    }

    private val TZDB_MIGRATION_SET = arrayOf(
            TZDB_MIGRATION_0_5, TZDB_MIGRATION_5_6, TZDB_MIGRATION_6_7
    )

    suspend fun getTimeZone(latitude: Double, longitude: Double): String? {
        if (latitude != 0.0 && longitude != 0.0) {
            AnalyticsLogger.logEvent("TZDBCache: querying")
            // Initialize db if it hasn't been already
            if (tzDB == null) {
                val context = SimpleLibrary.getInstance().appContext
                tzDB = Room.databaseBuilder(context,
                        TZDatabase::class.java, "tzdb.db")
                        .addMigrations(*TZDB_MIGRATION_SET)
                        .fallbackToDestructiveMigration()
                        .build()
            }

            // Search db if result already exists
            val dbResult = tzDB?.tzdbDAO()?.getTimeZoneData(latitude, longitude)

            if (!dbResult.isNullOrBlank())
                return dbResult

            // Search tz lookup
            val result = TimeZoneProvider().getTimeZone(latitude, longitude)

            if (!result.isNullOrBlank()) {
                // Cache result
                GlobalScope.launch(Dispatchers.IO) {
                    val tzdb = TZDB(latitude, longitude, result)
                    tzDB!!.tzdbDAO().insertTZData(tzdb)

                    // Run GC since tz lookup takes up a good chunk of memory
                    System.gc()
                }
            }

            return result
        }

        return "UTC"
    }
}